File: src/main/java/gq/genprog/autocrat/config/AutocratConfig.java
package gq.genprog.autocrat.config;

import net.minecraftforge.common.config.Config;

import static gq.genprog.autocrat.MetaKt.MOD_ID;

/**
 * Written by @offbeatwitch.
 * Licensed under MIT.
 */
@Config(modid = MOD_ID)
public class AutocratConfig {
	@Config.Comment({"Controls which Autocrat modules are enabled.",
			"Certain modules may force-disable themselves if they detect a mod with conflicting functionality."})
	public static Modules modules = new Modules();
	public static ClaimsModule claims = new ClaimsModule();
	public static SleepModule sleepVote = new SleepModule();
	public static BackupsModule backups = new BackupsModule();

	public static class Modules {
		@Config.Comment("Enable the claims modules (factions, claims)")
		public boolean claims = true;

		@Config.Comment("Enable the fancy names module (auto colour names, nicknames)")
		public boolean fancyNames = true;

		@Config.Comment("Enable the sleep-vote module. Disables itself if Quark or Morpheus are installed.")
		public boolean sleepVote = true;

		@Config.Comment("Enable the admin module (mod-mode)")
		public boolean admin = true;

		@Config.Comment("Enables the simple home module (ender pearl at feet)")
		public boolean simpleHome = true;

		@Config.Comment("Enables the command-based homes module (/home, /sethome). [WIP]")
		public boolean commandHome = false;
	}

	public static class ClaimsModule {
		@Config.Comment("Maximum number of chunks a group can claim.")
		public int maxClaimedChunks = 50;
	}

	public static class SleepModule {
		@Config.Comment("Controls the percentage of sleeping players required to skip to day.")
		public int threshold = 45;
	}

	public static class BackupsModule {
		@Config.Comment("Set the maximum size that can be restored at once via /bk restore.")
		public int maxRestoreSize = 524288;
	}
}

File: src/main/kotlin/gq/genprog/autocrat/config/AutocratConfig.kt
package gq.genprog.autocrat.config

import com.electronwill.nightconfig.core.UnmodifiableConfig
import net.minecraftforge.common.ForgeConfigSpec

object AutocratConfig {
    val spec = ForgeConfigSpec.Builder().apply {
        comment("Controls which Autocrat modules are enabled.",
                "Certain modules may force-disable themselves if they detect a mod with conflicting functionality.")
        push("modules").apply {
            comment("Enable the claims modules (factions, claims)")
            define("claims", true)
            
            comment("Enable the fancy names module (auto colour names, nicknames)")
            define("fancyNames", true)

            comment("Enable the sleep-vote module. Disables itself if Quark or Morpheus are installed.")
            define("sleepVote", true)

            comment("Enable the admin module (mod-mode)")
            define("admin", true)

            comment("Enables the simple home module (ender pearl at feet)")
            define("simpleHome", true)

            comment("Enables the command-based homes module (/home, /sethome). [WIP]")
            define("commandHome", false)
        }.pop()

        push("general").apply {
            comment("List of disabled commands by primary alias.")
            defineList("disabledCommands", listOf<String>()) { _ -> true }
        }.pop()

        push("claims").apply {
            comment("Maximum number of chunks a group can claim.")
            defineInRange("maxClaimedChunks", 50, 0, 1024)
        }.pop()

        push("sleep").apply {
            comment("Controls the percentage of sleeping players required to skip to day.")
            defineInRange("threshold", 40, 0, 100)
        }.pop()
    }.build()

    fun get() = ConfigAccess(spec.values)

    class ConfigAccess(val cfg: UnmodifiableConfig) {
        operator fun <T> get(path: String): T = cfg.get<ForgeConfigSpec.ConfigValue<T>>(path).get()
        fun getInt(path: String): Int = cfg.getInt(path)
    }

    fun <T> get(path: String): T = get()[path]
}